Use java.net.http.HttpClient instead of java.net.Http(s)URLConnection

translatenix opened the pull request:
Moving to java.net.http.HttpClient brings many benefits, including HTTP/2 support and the ability to make asynchronous requests.

Major additions and changes:
- Introduce a lightweight org.pkl.core.http.HttpClient API. This keeps some flexibility and allows to enforce behavior such as setting the User-Agent header.
- Provide an implementation that delegates to java.net.http.HttpClient.
- Use HttpClient for all HTTP(s) requests across the codebase. This required adding an HttpClient parameter to constructors and factory methods of multiple classes, some of which are public APIs.
- Manage CA certificates per HTTP client instead of per JVM. This makes it unnecessary to set JVM-wide system/security properties and default SSLSocketFactory's.

Each HTTP client maintains its own connection pool and SSLContext. For efficiency reasons, I've tried to reuse clients whenever feasible. To avoid memory leaks, clients are not stored in static fields.

HTTP clients are expensive to create. For this reason, EvaluatorBuilder defaults to a "lazy" client that creates the underlying java.net.http.HttpClient on the first send (which may never happen).

Fixes #157.

Note: This PR allows to fix the "flaky PackageServer tests" problem in a principled way. Because all HTTP requests are now routed through HttpClient, PackageServer can bind to a dynamic port, and HttpClient can rewrite requests to use that port in test mode. I've tested this approach on a local branch, and it's the first time I can get "gw clean build" to pass without "connection refused" errors.

bioball commented:
Thanks! I haven't reviewed this yet, but this is erroring in our CI build, for test `org.pkl.executor.EmbeddedExecutorTest`:

```
org.pkl.executor.ExecutorException: 
–– Pkl Error ––
Exception when making request `GET https://localhost:12110/birds@0.5.0`:
PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target

4 | import "package://localhost:12110/birds@0.5.0#/Bird.pkl"
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
at MyModule#Bird (file:///tmp/junit15463112887516922814/test.pkl)

6 | chirpy = new Bird { name = "Chirpy"; favoriteFruit { name = "Orange" } }
                 ^^^^
at MyModule#chirpy (file:///tmp/junit15463112887516922814/test.pkl)

106 | text = renderer.renderDocument(value)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
at pkl.base#Module.output.text (https://github.com/apple/pkl/blob/7a26325/stdlib/base.pkl#L106)
```

translatenix commented:
Unfortunately I can't reproduce locally. Is it possible to trigger PR builds automatically for contributors? Otherwise the feedback loop is too long.

bioball commented:
Re: CI: yeah, I realize it's kind of painful. We'll need to think through how to make this experience better for ya'll.

translatenix commented:
I found and fixed the problem. It wasn't showing locally because the test was using cache dir ~/.pkl/cache.

The fix makes breaking changes to spi.v1, which probably isn't a good idea. Let me know if you'd like me to do spi.v2 instead.

holzensp commented on pkl-core/src/test/kotlin/org/pkl/core/http/HttpClientTest.kt:
```diff
@@ -0,0 +1,141 @@
+  
+  @Test
+  fun `can load certificates from default location`(@TempDir userHome: Path) {
+    val certFile = userHome.resolve(".pkl")
```
I see the intent here, but I feel `userHome` is misleading naming in this test

translatenix replied:
Changed to `tempDir`.

holzensp reviewed (commented):
This looks really, really good... As to the API change; I _believe_ the incompatible changes you're making are all _additional_ fields in constructors and such (did I miss it?). Maybe it can stay `v1` if you add additional constructors that retain the old behaviour and just fill in reasonable defaults for the HttpClient in those cases.

Thoughts, @bioball?

translatenix commented on pkl-commons-cli/src/main/kotlin/org/pkl/commons/cli/commands/BaseOptions.kt:
```diff
@@ -142,28 +130,11 @@ class BaseOptions : OptionGroup() {
         names = arrayOf("--ca-certificates"),
         metavar = "<path>",
-        help = "Replaces the built-in CA certificates with the provided certificate file."
+        help = "Trust CA certificates from the provided file."
```
I chose this wording because this option replaces both ~/.pkl/cacerts and built-in certificates, which seemed too long to explain here. How about "Only trust CA certificates from the provided file(s)."? (The option is repeatable.)

bioball replied:
New wording LGTM

translatenix commented on pkl-cli/src/test/kotlin/org/pkl/cli/CliEvaluatorTest.kt:
```diff
@@ -1158,6 +1158,7 @@ result = someLib.x
   }
 
   @Test
+  @Disabled("flaky because CliEvaluator falls back to ~/.pkl/cacerts if no certs are given")
```
Fixed in a later commit. I've noticed that several CLI tests use default settings and hence use ~/.pkl/cacerts and ~/.pkl/cache, which makes them non-deterministic.

translatenix commented on pkl-commons-cli/src/main/kotlin/org/pkl/commons/cli/CliBaseOptions.kt:
```diff
@@ -167,4 +168,26 @@ data class CliBaseOptions(
+    } else {
+      for (file in normalizedCaCertificates) builder.addCertificates(file)
+    }
+    // Lazy building significantly reduces execution time of commands that do minimal work.
+    // However, it means that HTTP client initialization errors won't surface until an HTTP request
+    // is
+    // made. A middleground would be to only build lazily if built-in certificates are used.
```
The formatting is due to ktfmt, which seems to produce worse result than the Java formatter.

translatenix commented on pkl-core/src/main/java/org/pkl/core/http/RequestRewritingClient.java:
```diff
@@ -0,0 +1,96 @@
+    if (!closed.getAndSet(true)) delegate.close();
+  }
+
+  // Based on JDK 17's implementation of HttpRequest.newBuilder(HttpRequest, filter).
```
It means that the implementation can be simplified once JDK 17 is available.

translatenix replied:
It rewrites (modifies) the `HttpRequest`. From the Javadoc of `HttpClient.send`:

> If the request does not specify a timeout, the client's
> request timeout is used. If the request does not specify
> a preferred HTTP version, HTTP/2 is used. The request's
> User-Agent header is set to the client's User-Agent
>  header.

My `package-server` PR also rewrites the port if necessary.

translatenix commented on pkl-core/src/main/java/org/pkl/core/http/LazyHttpClient.java:
```diff
@@ -0,0 +1,77 @@
+import javax.annotation.concurrent.ThreadSafe;
+
+@ThreadSafe
+class LazyHttpClient implements HttpClient {
```
I can add some docs, though these are internal (package-private) classes, and the details are right in the code (same file).

translatenix replied:
Alright. I'll document internal classes if you promise to document public APIs such as `PackageResolver`. :-)

translatenix commented on pkl-core/src/main/java/org/pkl/core/module/ModuleKeys.java:
```diff
@@ -494,7 +509,7 @@ public ResolvedModuleKey resolve(SecurityManager securityManager)
         securityManager.checkResolveModule(redirected);
         var text = IoUtils.readString(stream);
-        return ResolvedModuleKeys.virtual(this, uri, text, true);
+        return ResolvedModuleKeys.virtual(this, redirected, text, true);
```
I pondered on this for a while because I had to make the same decision a few lines above. I concluded that the use of `uri` here is a bug. It means that `ModuleCache.modulesByResolvedUri` will cache the module under its unresolved URI, which is exactly what `ModuleCache.modulesByOriginalUri` does. So `modulesByResolvedUri` no longer serves any purpose.

bioball replied:
"resolve" here means resolving the in-language URI to the actual URI used for fetching a module's contents.

For example, an in-language URI might be `projectpackage://pkg.pkl-lang.com/foo@1.0.0#/foo.kl`, whereas the resolved URI might be `file:///path/to/foo.pkl`. It's useful here because the same actual file can be imported using two _different_ URIs in language, but actually are both the same file and we don't need to load it twice.

> It means that ModuleCache.modulesByResolvedUri will cache the module under its unresolved URI, which is exactly what ModuleCache.modulesByOriginalUri does. So modulesByResolvedUri no longer serves any purpose

There's plenty of module keys that have the same entries in both `modulesByResolvedUri` and `modulesByOriginalUri` (e.g. `file:` modules). It's expected for things like this.

translatenix commented on pkl-commons-cli/src/main/kotlin/org/pkl/commons/cli/CliBaseOptions.kt:
```diff
@@ -167,4 +168,26 @@ data class CliBaseOptions(
+  /**
+   * The HTTP client shared between CLI commands created with this [CliBaseOptions] instance.
+   *
+   * To release the resources held by the HTTP client in a timely manner, call its `close()` method.
```
I didn't change `CliBaseOptions` to implement `AutoCloseable`. Usually, an options object shouldn't need to be closed. Adding `CliBaseOptions.httpClient` felt a bit smelly, but it seemed like a price worth paying for being able to share `HttpClient` resources between commands. (I couldn't find another way without a major redesign.) I documented calling `CliBaseOptions.httpClient.close()` for the edge cases where it might matter.

Let me know if you want me to change `CliBaseOptions` to implement `AutoCloseable`. Kotlin doesn't have try-with-resources, but it does have a `Closeable.use` extension method.

translatenix replied:
> Probably makes more sense to either accept the options on HttpClient.Builder directly, or accept HttpClient.Builder as an option as a whole.

Sorry I don't understand what you're proposing. Can you elaborate?
I considered passing `HttpClient` to `CliCommand`, but then `CliBaseOptions.caCertificates` is no longer meaningful. (It's `HttpClient` that needs to be configured with the certificates.)

bioball replied:
Sorry, so, one suggestion is to add the builder methods on `HttpClient.Builder` to `EvaluatorBuilder`:

```java
class EvaluatorBuilder {
  public EvaluatorBuilder addCertificates(Path file) { /* etc */ }

  public EvaluatorBuilder addCertificates(URI file) { /* etc */ }
```

Another option is to accept an `HttpClient.Builder` as an option in `EvaluatorBuilder:

```java
class EvaluatorBuilder {
  public HttpClient.Builder setHttpClientBuilder(HttpClient.Builder builder) { /* etc */ }
}
```

translatenix replied:
Currently we have `EvaluatorBuilder.setHttpClient()`. Your proposals might make EvaluatorBuilder more convenient to use if the default HTTP client isn't good enough. But I don't see how any of this relates to the `CliBaseOptions.httpClient` discussion, and it's not clear to me if and what change you want there.

translatenix commented on pkl-cli/pkl-cli.gradle.kts:
```diff
@@ -156,7 +156,7 @@ fun Exec.configureExecutable(isEnabled: Boolean, outputFile: File, extraArgs: Li
         ,"-H:IncludeResources=org/pkl/core/stdlib/.*\\.pkl"
         ,"-H:IncludeResources=org/jline/utils/.*"
-        ,"-H:IncludeResources=org/pkl/commons/cli/commands/IncludedCARoots.pem"
+        ,"-H:IncludeResources=org/pkl/core/http/IncludedCARoots.pem"
```
The resource is now exclusively used by `pkl.core.http.HttpClientBuilder`, so moving it was natural. I believe it is also necessary because core doesn't depend on commons-cli.

translatenix commented on pkl-core/src/main/java/org/pkl/core/module/ResolvedModuleKeys.java:
```diff
@@ -102,6 +106,14 @@ public URI getUri() {
 
     @Override
     public String loadSource() throws IOException {
+      if (HttpUtils.isHttpUrl(url)) {
+        var httpClient = VmContext.get(null).getHttpClient();
+        var request = HttpRequest.newBuilder(uri).build();
+        var response = httpClient.send(request, BodyHandlers.ofString());
+        HttpUtils.checkHasStatusCode200(response);
+        return response.body();
+      }
```
There are too many differences to share with `ResourceReaders`, and I can't spot any other repetitions.

translatenix commented on pkl-core/src/main/java/org/pkl/core/resource/ResourceReaders.java:
```diff
@@ -291,6 +294,15 @@ public boolean isGlobbable() {
   private abstract static class UrlResource implements ResourceReader {
     @Override
     public Optional<Object> read(URI uri) throws IOException {
+      if (HttpUtils.isHttpUrl(uri)) {
+        var httpClient = VmContext.get(null).getHttpClient();
+        var request = HttpRequest.newBuilder(uri).build();
+        var response = httpClient.send(request, BodyHandlers.ofByteArray());
+        if (response.statusCode() == 404) return Optional.empty();
+        HttpUtils.checkHasStatusCode200(response);
+        return Optional.of(new Resource(uri, response.body()));
+      }
```
There are too many differences to share with `ResolvedModuleKeys`, and I can't spot any other repetitions.

bioball commented on pkl-core/src/main/java/org/pkl/core/service/ExecutorSpiImpl.java:
```diff
@@ -98,4 +126,53 @@ public String evaluatePath(Path modulePath, ExecutorSpiOptions options) {
+  }
+
+  private static final class HttpClientKey {
+    final Set<Path> certificateFiles;
+
+    @SuppressWarnings("UrlHashCode") // only intended to be used for class path URLs
+    final Set<URL> certificateUrls;
```
Switch to `List<URL>`

translatenix replied:
I decided to accept `URL` certificate files in APIs (only) to support loading certificate files from the class path. If we're genuinely concerned about potential side effects of `URL.equals`, I'd suggest to immediately reject URLs that aren't `jar:` or `file:` at API boundaries. Additionally we could use `URI` in APIs and/or internally and only convert to `URL` when accessing the (`jar:` or `file:`) resource. This is probably a worthwhile change; I just wasn't sure if it's necessary.

translatenix replied:
> There's plenty of module keys that have the same entries in both modulesByResolvedUri and modulesByOriginalUri (e.g. file: modules). It's expected for things like this.

OK, I'll revert to the current behavior. That said, conceptually it feels wrong that Pkl treats A and B where A redirects to B as different modules. And if it does, it would seem more consistent to forego a security check of B when accessing A.

bioball replied:
Not totally sure what you mean, but, it's important that these two are different modules, even if the first URI redirects to the second URI. It would be hard for users to predict the language's behavior otherwise.

```
import "https://example.com/foo.pkl"
import "https://example.com/bar.pkl"
```

translatenix replied:
Moved discussion to #254

bioball commented on pkl-executor/src/main/java/org/pkl/executor/spi/v1/ExecutorSpiOptions.java:
```diff
@@ -83,6 +91,14 @@ public List<Path> getModulePath() {
+  public List<URL> getCertificateUrls() {
+    return certificateUrls;
+  }
+
```
I think I'd prefer to not change this.

If we add this, we'll need to introduce `v2`. And, for the executor use-case, you likely want Pkl to trust the same certs that the JVM already trusts anyways. It doesn't seem compelling enough to add a new knob here just yet.

I think let's have `ExecutorSpiImpl` use an http client that uses the default `SSLContext`. And in `EmbeddedExecutorTest`, we can use the good ol' `CertificateUtils.setupAllX509CertificatesGlobally` to get those tests to pass. Except move that class into `pkl-commons-test`.

translatenix replied:
I don't mind not supporting certificates in Executor. But I really feel that `CertificateUtils.setupAllX509CertificatesGlobally` needs to go. It's a gross hack that's asking for trouble (might get used elsewhere, might interfere with other tests, code duplication with JdkHttpClient, etc.).

PS: My `package-server` PR also breaks `spi.v1`. Perhaps it's time for `spi.v2` anyways?

bioball replied:
By moving `CertificateUtils.setupAllX509CertificatesGlobally` into `pkl-commons-test`, we ensure that it's only used for testing purposes. And likely, the embedded executor test will be the only place that uses it, so we can even just inline its contents into that test itself.

> PS: My package-server PR also breaks spi.v1. Perhaps it's time for spi.v2 anyways?

Hm, I haven't reviewed that one yet. I'll need to take a look. Yeah if we do introduce spi.v2, might as well keep these options.

translatenix replied:
> By moving CertificateUtils.setupAllX509CertificatesGlobally into pkl-commons-test, we ensure that it's only used for testing purposes.

I think it's only a matter of time until this will once again cause flaky tests. Mutating global state is best avoided. I'd be more than happy to implement `spi.v2` if that's what it takes.

bioball reviewed (changes requested):
Amazing work! This is looking really great.

Some things that need changes (details are in my comments)

* Move built-in CA certs back to `pkl-commons`
* Let's not touch the executor API
* `EvaluatorBuilder.preconfigured()` should trust the JVM's certs

Also: what was the thinking around introducing `org.pkl.core.http.HttpClient`? 

I see the benefits that it brings (`DummyHttpClient` is quite nice). But, it also puts the burden on us to expose all the settings someone might want, that are already in `java.net.http.HttpClient.Builder`. Also, maybe users already have their own `java.net.HttpClient` that they want to use and share with the rest of their application?

translatenix commented on pkl-commons-cli/src/main/kotlin/org/pkl/commons/cli/CliBaseOptions.kt:
```diff
@@ -113,15 +114,15 @@ data class CliBaseOptions(
    *
-   * If not empty, this determines the CA root certs used for all HTTPS requests. Warning: this
-   * affects the whole Java runtime, not just the Pkl API!
+   * If [caCertificates] is the empty list, the certificate files in `~/.pkl/cacerts/` are used. If
```
My aim was be clear about empty list vs. empty file.

bioball replied:
Ah, I see. Yeah, that's fine as is then.

translatenix commented on pkl-core/src/main/java/org/pkl/core/http/HttpClient.java:
```diff
@@ -0,0 +1,165 @@
+    /**
+     * Sets the {@code User-Agent} header.
+     *
+     * <p>Defaults to {@code "Pkl/$version ($os $flavor)"}.
```
> We intentionally kept this separate from pkl-core because on the Java side, we wanted our HTTP requests to defer to the host application's settings (e.g. trust the same CA roots).

Defaulting to the JDK's CA roots was my original goal, but I couldn't find a way to make it work. This is best discussed separately. 

In any case, if we want to support `HttpClient.Builder.addBuiltInCertificates` and `HttpClient.Builder.addDefaultCliCertificates`, which makes a lot of sense to me, then `IncludedCARoots.pem` must be accessible from `pkl-core`. As a consequence, it cannot reside in `pkl-commons` or `pkl-commons-cli` (its original location) because neither is a dependency of `pkl-core`.

bioball replied:
> I don't see any downsides of having this file in pkl-core.

The main downside is filesize; the file is 213Kb. All of our JVM users end up with this file in their classpath and likely none of them will need it.

translatenix replied:
How about putting the built-in certs into a `pkl-certs` JAR? Then everyone who wants them can have them, without pulling in any other JAR. (This isn't even limited to JVM use cases because a JAR is just a ZIP.) And everyone else can stay away from them. And we can stop guessing.

bioball replied:
That sounds fine with me!

translatenix replied:
Should this JAR be versioned with Pkl? I'm asking because it will rarely change. And should it actually be a ZIP file, i.e., have a `.zip` extension? `ClassLoader.getResource(pathToZipFile)` is officially supported.

translatenix replied:
> <p>Defaults to {@code "Pkl/$version ($os; $flavor)"}.
fixed

bioball replied:
> Should this JAR be versioned with Pkl? I'm asking because it will rarely change. And should it actually be a ZIP file, i.e., have a .zip extension? ClassLoader.getResource(pathToZipFile) is officially supported.

I was thinking it would be its own project (pkl-certs, sibling of pkl-cli, etc). And it would publish its own jar to Maven. So yeah, versioned with Pkl.

What were you thinking here?

translatenix replied:
For now I did `pkl-certs` because this was easiest. But this *could* be a separate GitHub project with initial version 1. And it *could* publish a `.zip` file to Maven Central. I don't really mind but wanted to mention the alternatives.

bioball replied:
Ah I see. Let's keep it here.

translatenix commented on pkl-core/src/main/java/org/pkl/core/http/JdkHttpClient.java:
```diff
@@ -0,0 +1,194 @@
+  private static final MethodHandle closeMethod;
+
+  static {
+    var methodType = MethodType.methodType(void.class, java.net.http.HttpClient.class);
+    MethodHandle result;
+    try {
+      //noinspection JavaLangInvokeHandleSignature
+      result =
+          MethodHandles.publicLookup()
+              .findVirtual(java.net.http.HttpClient.class, "close", methodType);
+    } catch (NoSuchMethodException | IllegalAccessException e) {
+      // use no-op close method
+      result = MethodHandles.empty(methodType);
+    }
```
See comment in line 59. What else should I add?

translatenix commented on pkl-commons-cli/src/main/kotlin/org/pkl/commons/cli/CliBaseOptions.kt:
```diff
@@ -167,4 +168,26 @@ data class CliBaseOptions(
+   * To release the resources held by the HTTP client in a timely manner, call its `close()` method.
+   */
+  val httpClient: HttpClient by lazy {
+    val builder = HttpClient.builder()
+    if (normalizedCaCertificates.isEmpty()) {
+      builder.addDefaultCliCertificates()
+    } else {
+      for (file in normalizedCaCertificates) builder.addCertificates(file)
+    }
+    // Lazy building significantly reduces execution time of commands that do minimal work.
+    // However, it means that HTTP client initialization errors won't surface until an HTTP request
+    // is
+    // made. A middleground would be to only build lazily if built-in certificates are used.
+    builder.buildLazily()
```
> Move built-in CA certs back to pkl-commons

They never were in pkl-commons. Details elsewhere.

> Let's not touch the executor API

Further discussed elsewhere.

> EvaluatorBuilder.preconfigured() should trust the JVM's certs

I think the best I can do is to use `SSLContext.getDefault()`, which is more or less what you're asking for.

> Also: what was the thinking around introducing org.pkl.core.http.HttpClient?

At first I was using `java.net.http.HttpClient` directly. But baking this interface into Pkl APIs didn't feel right. Having our own abstraction has clear benefits:

* It puts us in control: We can remove what is irrelevant to Pkl, and add or change what is beneficial.
* A smaller interface is easier to implement and doesn't require throwing `UnsupportedOperationException` for all the methods that a particular implementation may not be able or willing to support. There are already 4 implementations, and chances are there will be more.
* For the same reason, it will be easier to support alternative implementations (Apache, Netty, etc.) if the need ever arises. While `java.net.http.HttpClient` is a clear step up from `HttpURLConnection`, it will always remain less capable than third-party libraries. For example, Netty supports a safer TLS implementation.

> But, it also puts the burden on us to expose all the settings someone might want, that are already in java.net.http.HttpClient.Builder.
> Also, maybe users already have their own java.net.HttpClient that they want to use and share with the rest of their application?

We could easily support this with something like an `HttpClient from(java.net.http.HttpClient)` factory method.

I didn't want to go so far as to have our own `HttpRequest`, `HttpResponse`, etc. Having "just" our own `HttpClient` felt like the sweet spot for Pkl.

translatenix replied:
> `apply` is a nice way to reduce a bit of boilerplate.

Changed to `with(HttpClient.builder()) {...}`.

translatenix commented:
@holzensp @bioball I think I've addressed all feedback (see recent commits) except for the following open questions:

1. `pkl-certs` details (see that thread)
2. `org.pkl.executor.Executor` details (see that thread)
3. `CliBaseOptions.httpClient` details (see that thread)

translatenix commented:
@holzensp @bioball I'm blocked on your feedback.
I'd love to land this PR and #227 ASAP to stabilize the build/tests. It will make everything easier for everyone.

> I think I've addressed all feedback (see recent commits) except for the following open questions:
> 1. pkl-certs details (see that thread)
> 2. org.pkl.executor.Executor details (see that thread)
> 3. CliBaseOptions.httpClient details (see that thread)

bioball commented:
@translatenix: ack; I'll get back to you by end of day tomorrow. Thanks again for your work here, and bear with us as we work through all the PRs (including our own).

translatenix commented:
I've now implemented spi.v2 to support certificate files (and `testPort` in #227 ) without introducing breaking changes.
I've also made sure that pkl-certs gets published.

This is ready to be merged from my side.
I'd prefer to defer non-essential improvements to this PR until the build has been stabilized (#227 ).
Stabilizing the build should be the highest priority.

bioball commented on pkl-core/src/main/java/org/pkl/core/service/ExecutorSpiImpl.java:
```diff
@@ -80,22 +173,64 @@ public String evaluatePath(Path modulePath, ExecutorSpiOptions options) {
+          }
+          // If the above didn't add any certificates,
+          // builder will use the JVM's default SSL context.
+          return builder.build();
```
Lazy?

```suggestion
          return builder.buildLazily();
```

translatenix replied:
fixed

bioball commented on pkl-executor/src/main/java/org/pkl/executor/spi/v2/ExecutorSpiOptions2.java:
```diff
@@ -0,0 +1,120 @@
+import java.util.List;
+import java.util.Map;
+
+public class ExecutorSpiOptions2 {
```
For just adding more options, let's make this part of the v1 SPI, but just as an additional class that extends `ExecutorSpiOptions`. This simplifies the SPI stuff quite a bit.

I've prepared a diff that demonstrates the changes that I'm talking about here (not fully tested):

https://gist.github.com/bioball/c1399aabc99e44103408085f0d03133a

translatenix replied:
I don’t think this works. Classes in the spi package are loaded by the class loader that loads pkl-executor classes (see PklDistributionClassloader.loadClass). If you have an old pkl-executor that only has v1.ExecutorSpiOptions, and a newer pkl-core that tries to instantiate v1.ExecutorSpiOptions2, you’ll get a NoClassDefFoundError.

SPI versioning is tricky business. I think it’s best to keep everything strictly separated even if it means some code repetition. (I tried making v2.ExecutorSpiOptions2 extend v1.ExecutorSpiOptions, which I *think* should work, but it didn’t help all that much, and I didn’t feel it was worth the risk and entanglement between versions.) Note that going from 2 to 3 versions will be easier than going from 1 to 2 because I more or less added the infrastructure needed to support n versions.

translatenix replied:
> We need to detail this somewhere, but--PackageResolver isn't a public API.

Currently it is, by definition:
* `EvaluatorBuilder` is a public API. It has a public method `setProjectDependencies(org.pkl.core.project.DeclaredDependencies)` -> `org.pkl.core.project` is a public API.
* `org.pkl.core.project.ProjectDependenciesResolver` has a public constructor that takes `org.pkl.core.packages.PackageResolver` -> `org.pkl.core.packages` is a public API.
* `PackageResolver.getDependencyMetadataAndComputeChecksum` returns `org.pkl.core.util.Pair` -> `org.pkl.core.util` is a public API.

Formalizing the public API by adding support for JPMS (https://github.com/apple/pkl/issues/42) will be a good exercise. It will also likely require breaking API changes. It's something I'd be interested to work on.

translatenix replied:
The way I know this from other projects is that either side should be able to update without breaking the other to avoid a lock-in. If there’s a case that doesn’t work, I’m happy to fix it (and add test coverage) after the build has been stabilized.

bioball replied:
I'd be happy to merge this PR if it didn't include these SPI changes.

Can you omit the SPI changes from this PR for now? And let's address that in a follow-up PR.

In #227, feel free to add `@Disabled` to tests `evaluate a module that loads a package` and `evaluate a project dependency` so we can stabilize the build without needing to bring in SPI changes.

translatenix replied:
You mean to neither add spi.v2 nor change spi.v1 to pass certificates? Then I’ll also need to disable some tests in this PR.

bioball replied:
Yeah; no SPI changes for this PR. And yeah, LGTM to disable those tests in this PR too.

translatenix replied:
How about accepting breaking spi.v1 changes for now. That requires far fewer changes, and this needs to be fixed anyway.

bioball replied:
I know this is higher friction for you, but, I'd prefer to not merge something into main that isn't releasable. If it feels like too much churn for you, I can prepare a version of your PR that excludes the SPI changes (and same with #227)

bioball commented on pkl-certs/pkl-certs.gradle.kts:
```diff
@@ -0,0 +1,19 @@
+        url.set("https://github.com/apple/pkl/tree/main/pkl-certs")
+        description.set("""
+          Pkl's built-in CA certificates. 
+          Used by pkl-cli and optionally supported by pkl-core.")
```
Also used by pkl-doc, pkl-codegen-java, pkl-codegen-kotlin, etc.

```suggestion
          Used by Pkl's CLIs, and optionally supported by pkl-core.")

```

translatenix replied:
fixed

bioball reviewed (changes requested):
No major issues. Thanks again for your contribution!

One blocking issue is to simplify the SPI v2 options (see comment below)

translatenix commented on pkl-core/src/main/java/org/pkl/core/http/HttpClientBuilder.java:
```diff
@@ -0,0 +1,135 @@
+  }
+
+  public HttpClient.Builder addDefaultCliCertificates() {
+    var directory = userHome.resolve(".pkl").resolve("cacerts");
```
can't use this here because `userHome` is configurable to enable testing

bioball replied:
Ah, I see. How about:

```
HttpClientBuilder() {
  this(IoUtils.getPklHomeDir());
}

HttpClientBuilder(Path pklHomeDir) {
  this.pklHomeDir = pklHomeDir
}
```

translatenix commented on pkl-core/src/main/java/org/pkl/core/http/HttpClient.java:
```diff
@@ -0,0 +1,182 @@
+ *
+ * <p>To create a new HTTP client, use a {@linkplain #builder() builder}. To send {@linkplain
+ * HttpRequest requests} and retrieve their {@linkplain HttpResponse responses}, use {@link #send}.
+ * To release resources held by the client in a timely manner, use {@link #close}.
```
> I'm proposing that both here, and in evaluator builder, we provide knobs for addCertificates, and possibly any other options on HttpClient.Builder.

There's already a `CliBaseOptions.caCertificates` constructor parameter. There is currently no builder for `CliBaseOptions`.
Adding convenience methods to `EvaluatorBuilder` is possible, but let's defer this. I could do this in #197 , which already adds a couple of convenience methods to `EvaluatorBuilder`.

> the wording is slightly vague

It's because the underlying JDK client only offers weak guarantees, and at least connections will eventually expire even if `close()` isn't called. (Before JDK 21 there isn't even a `close()` method.) But I'll remove this nevertheless.

translatenix commented:
I've addressed everything I could in my latest commit (as per my comments). Once you're ready to merge, please squash before merging (or ask me to squash).

translatenix commented on pkl-commons-cli/src/main/kotlin/org/pkl/commons/cli/CliCommand.kt:
```diff
@@ -158,6 +152,10 @@ abstract class CliCommand(protected val cliOptions: CliBaseOptions) {
     )
   }
 
+  // share HTTP client with other commands with the same cliOptions
```
It’s the same *instance* that’s shared, which is especially important for HttpClient and not generally the case for other protected members.

bioball commented:
> I've addressed everything I could in my latest commit (as per my comments). Once you're ready to merge, please squash before merging (or ask me to squash).

We will definitely squash when merging

translatenix commented:
Instead of backing out all SPI changes, I switched to your suggested implementation approach.
To gain some confidence in this approach, I massively improved `EmbeddedExecutorTest`.
Details in the commit message.

If you prefer to back out all SPI changes, I'll take you up on your offer to take over from here.

bioball commented on pkl-executor/pkl-executor.gradle.kts:
```diff
@@ -49,7 +51,17 @@ sourceSets {
+  src("https://repo1.maven.org/maven2/org/pkl-lang/pkl-config-java-all/0.25.0/pkl-config-java-all-0.25.0.jar")
+  dest("build/download/pkl-config-java-all-0.25.0.jar")
+  doFirst { file("build/download").mkdirs() }
+}
```
[nit] this is a little nicer:

```kotlin
val pklHistoricalDistributions: Configuration by configurations.creating

dependencies {
  @Suppress("UnstableApiUsage")
  pklHistoricalDistributions(libs.pklConfigJavaAll025)
}

val prepareHistoricalDistributions by tasks.registering(Copy::class) {
  from(pklHistoricalDistributions)
  into(layout.buildDirectory.dir("pklDistributions"))
}

val prepareTest by tasks.registering {
  dependsOn(prepareHistoricalDistributions, pklDistribution)
}
```

translatenix replied:
Ack. I went with a solution that doesn't copy. Switching to IntelliJ's Gradle test runner would simplify matters, but it's 10x slower than the already super slow platform test runner on WSL.

bioball replied:
Can we keep the copy solution?

This piece of code seems too brittle:

```java
        System.getProperty("user.home").toPath()
          .resolve(".gradle/caches/modules-2/files-2.1/org.pkl-lang/pkl-config-java-all/" +
            "0.25.0/e9451dda554f1659e49ff5bdd30accd26be7bf0f/pkl-config-java-all-0.25.0.jar")
```

translatenix replied:
It's supposed to be deterministic and only used when running tests with IntelliJ's built-in test runner. But I can do a symlink instead, which also avoids a copy after every clean.

bioball replied:
Using IJ's JUnit run configurations have always been the better experience for me, over the Gradle run configurations. I actually like `gw prepTest` as a way to bridge the gap for us here.

bioball commented on pkl-core/src/main/java/org/pkl/core/service/ExecutorSpiImpl.java:
```diff
@@ -98,4 +128,61 @@ public String evaluatePath(Path modulePath, ExecutorSpiOptions options) {
+  private HttpClient getOrCreateHttpClient(ExecutorSpiOptions options) {
+    List<Path> certificateFiles;
+    List<URI> certificateUris;
+    if (options instanceof ExecutorSpiOptions2) {
+      var options2 = (ExecutorSpiOptions2) options;
+      certificateFiles = options2.getCertificateFiles();
+      certificateUris = options2.getCertificateUris();
+    } else {
```
To handle if pkl-executor is older than this distribution:

```suggestion
    try {
      if (options instanceof ExecutorSpiOptions2) {
        var options2 = (ExecutorSpiOptions2) options;
        certificateFiles = options2.getCertificateFiles();
        certificateUris = options2.getCertificateUris();
      } else {
        certificateFiles = List.of();
        certificateUris = List.of();
      }
    } catch (NoClassDefFoundError e) {
```

And we can enable the test if the executor is older than the distribution version.

translatenix replied:
This is really testing the limits, and I can't actually get it to work. I think it's far safer to ship `pkl-executor` with `pkl-config-java-all` or some other distribution.

translatenix replied:
We can probably find a solution that won’t require including pkl-executor in distributions. Until then, we can leave the “older pkl-executor, newer distribution” test disabled.

bioball replied:
This change is passing for me:

```diff
diff --git a/pkl-core/src/main/java/org/pkl/core/service/ExecutorSpiImpl.java b/pkl-core/src/main/java/org/pkl/core/service/ExecutorSpiImpl.java
index 3af2df3eb..1c713cb76 100644
--- a/pkl-core/src/main/java/org/pkl/core/service/ExecutorSpiImpl.java
+++ b/pkl-core/src/main/java/org/pkl/core/service/ExecutorSpiImpl.java
@@ -132,11 +132,16 @@ public class ExecutorSpiImpl implements ExecutorSpi {
   private HttpClient getOrCreateHttpClient(ExecutorSpiOptions options) {
     List<Path> certificateFiles;
     List<URI> certificateUris;
-    if (options instanceof ExecutorSpiOptions2) {
-      var options2 = (ExecutorSpiOptions2) options;
-      certificateFiles = options2.getCertificateFiles();
-      certificateUris = options2.getCertificateUris();
-    } else {
+    try {
+      if (options instanceof ExecutorSpiOptions2) {
+        var options2 = (ExecutorSpiOptions2) options;
+        certificateFiles = options2.getCertificateFiles();
+        certificateUris = options2.getCertificateUris();
+      } else {
+        certificateFiles = List.of();
+        certificateUris = List.of();
+      }
+    } catch (NoClassDefFoundError e) {
       certificateFiles = List.of();
       certificateUris = List.of();
     }
diff --git a/pkl-executor/src/test/kotlin/org/pkl/executor/EmbeddedExecutorTest.kt b/pkl-executor/src/test/kotlin/org/pkl/executor/EmbeddedExecutorTest.kt
index 3658b3ead..46e014274 100644
--- a/pkl-executor/src/test/kotlin/org/pkl/executor/EmbeddedExecutorTest.kt
+++ b/pkl-executor/src/test/kotlin/org/pkl/executor/EmbeddedExecutorTest.kt
@@ -39,7 +39,7 @@ class EmbeddedExecutorTest {
 
         // This context has a pkl-executor version that is lower than the distribution version.
         // It can be enabled once there is a distribution that includes pkl-executor.
-        //ExecutionContext(executor1_2.value, ::convertToOptions1, "Options1, Executor1, Distribution2"),
+        ExecutionContext(executor1_2.value, ::convertToOptions1, "Options1, Executor1, Distribution2"),
 
         ExecutionContext(executor2_1.value, ::convertToOptions1, "Options1, Executor2, Distribution1"),
         ExecutionContext(executor2_1.value, ::convertToOptions2, "Options2, Executor2, Distribution1"),
```

translatenix replied:
Didn't work for me, but maybe IntelliJ didn't pick up my code change (I've been struggling with this a lot). I don't know if `NoClassDefFoundError` is guaranteed to be only thrown here though. It's a fairly extreme solution.

bioball reviewed (dismissed):
Some nits, but overall LGTM. Great work!

Passing over to @stackoverflow for review

stackoverflow commented on pkl-executor/src/test/kotlin/org/pkl/executor/EmbeddedExecutorTest.kt:
```diff
@@ -1,42 +1,159 @@
+    }
+
+    // A pkl-executor library that supports ExecutorSpiOptions up to v2
+    // and a Pkl distribution that supports ExecutorSpiOptions up to v.
```
```suggestion
    // and a Pkl distribution that supports ExecutorSpiOptions up to v2.
```

stackoverflow reviewed (approved):
Looks good. Left some comments.

translatenix commented on pkl-core/src/main/java/org/pkl/core/http/RequestRewritingClient.java:
```diff
@@ -0,0 +1,108 @@
+                  builder.GET();
+                  break;
+                case "DELETE":
+                  builder.DELETE();
```
When a builder and the getters of the object it builds aren't symmetric, copying becomes more difficult. Copied from JDK 17 code.

bioball commented:
@translatenix: are you planning on addressing the nits here? If so, I'll wait for them before merging. Otherwise, we can go ahead and merge, and we will submit a follow-up cleanup PR.
